### ISSUE TITLE
_cext.pyd depends on MSVCP140.dll, which is not distributed

### ISSUE THREAD

**Issue by LordAro** [nucleic/kiwi#187]
Version: 1.4.7

_cext.pyd still depends MSVCP140.dll, even though this was supposed to have been fixed by #179 (?)

```
$ python3.12 -m pip download kiwisolver
Collecting kiwisolver
  Using cached kiwisolver-1.4.7-cp312-cp312-win_amd64.whl.metadata (6.4 kB)
Using cached kiwisolver-1.4.7-cp312-cp312-win_amd64.whl (55 kB)
Saved c:\foo\kiwisolver-1.4.7-cp312-cp312-win_amd64.whl
Successfully downloaded kiwisolver

$ file kiwisolver-1.4.7-cp312-cp312-win_amd64.whl
kiwisolver-1.4.7-cp312-cp312-win_amd64.whl: Zip archive data, at least v2.0 to extract, compression method=deflate

$ unzip kiwisolver-1.4.7-cp312-cp312-win_amd64.whl
Archive:  kiwisolver-1.4.7-cp312-cp312-win_amd64.whl
  inflating: kiwisolver/__init__.py
  inflating: kiwisolver/_cext.cp312-win_amd64.pyd
  inflating: kiwisolver/_cext.pyi
  inflating: kiwisolver/exceptions.py
  inflating: kiwisolver/py.typed
  inflating: kiwisolver-1.4.7.dist-info/LICENSE
  inflating: kiwisolver-1.4.7.dist-info/METADATA
  inflating: kiwisolver-1.4.7.dist-info/WHEEL
  inflating: kiwisolver-1.4.7.dist-info/top_level.txt
  inflating: kiwisolver-1.4.7.dist-info/RECORD

$ objdump -p kiwisolver/_cext.cp312-win_amd64.pyd | grep DLL
        DLL
 vma:            Hint    Time      Forward  DLL       First
        DLL Name: MSVCP140.dll
        DLL Name: python312.dll
        DLL Name: KERNEL32.dll
        DLL Name: VCRUNTIME140_1.dll
        DLL Name: VCRUNTIME140.dll
        DLL Name: api-ms-win-crt-runtime-l1-1-0.dll
        DLL Name: api-ms-win-crt-stdio-l1-1-0.dll
        DLL Name: api-ms-win-crt-heap-l1-1-0.dll
```

Found by installing Python in [Windows Sandbox](https://learn.microsoft.com/en-us/windows/security/application-security/application-isolation/windows-sandbox/windows-sandbox-overview) and trying to import matplotlib. Matplotlib (3.9.2) does not depend on these external dependencies

<details>
<summary>Matplotlib deps</summary>

```
$ find matplotlib -name '*.pyd' -print -exec sh -c 'objdump -p {} | grep DLL' \;
matplotlib/backends/_backend_agg.cp312-win_amd64.pyd
        DLL
 vma:            Hint    Time      Forward  DLL       First
        DLL Name: api-ms-win-crt-runtime-l1-1-0.dll
        DLL Name: api-ms-win-crt-string-l1-1-0.dll
        DLL Name: api-ms-win-crt-math-l1-1-0.dll
        DLL Name: VCRUNTIME140.dll
        DLL Name: python312.dll
        DLL Name: api-ms-win-crt-heap-l1-1-0.dll
        DLL Name: VCRUNTIME140_1.dll
        DLL Name: KERNEL32.dll
matplotlib/backends/_tkagg.cp312-win_amd64.pyd
        DLL
 vma:            Hint    Time      Forward  DLL       First
        DLL Name: api-ms-win-crt-runtime-l1-1-0.dll
        DLL Name: api-ms-win-crt-heap-l1-1-0.dll
        DLL Name: api-ms-win-crt-string-l1-1-0.dll
        DLL Name: api-ms-win-crt-math-l1-1-0.dll
        DLL Name: VCRUNTIME140.dll
        DLL Name: COMCTL32.dll
        DLL Name: PSAPI.DLL
        DLL Name: python312.dll
        DLL Name: KERNEL32.dll
        DLL Name: VCRUNTIME140_1.dll
matplotlib/ft2font.cp312-win_amd64.pyd
        DLL
 vma:            Hint    Time      Forward  DLL       First
        DLL Name: api-ms-win-crt-runtime-l1-1-0.dll
        DLL Name: api-ms-win-crt-math-l1-1-0.dll
        DLL Name: api-ms-win-crt-heap-l1-1-0.dll
        DLL Name: api-ms-win-crt-stdio-l1-1-0.dll
        DLL Name: api-ms-win-crt-string-l1-1-0.dll
        DLL Name: api-ms-win-crt-locale-l1-1-0.dll
        DLL Name: VCRUNTIME140.dll
        DLL Name: python312.dll
        DLL Name: api-ms-win-crt-utility-l1-1-0.dll
        DLL Name: api-ms-win-crt-convert-l1-1-0.dll
        DLL Name: VCRUNTIME140_1.dll
        DLL Name: KERNEL32.dll
matplotlib/_c_internal_utils.cp312-win_amd64.pyd
        DLL
 vma:            Hint    Time      Forward  DLL       First
        DLL Name: api-ms-win-crt-runtime-l1-1-0.dll
        DLL Name: api-ms-win-crt-heap-l1-1-0.dll
        DLL Name: api-ms-win-crt-string-l1-1-0.dll
        DLL Name: api-ms-win-crt-math-l1-1-0.dll
        DLL Name: VCRUNTIME140.dll
        DLL Name: ole32.dll
        DLL Name: SHELL32.dll
        DLL Name: USER32.dll
        DLL Name: python312.dll
        DLL Name: KERNEL32.dll
        DLL Name: VCRUNTIME140_1.dll
matplotlib/_image.cp312-win_amd64.pyd
        DLL
 vma:            Hint    Time      Forward  DLL       First
        DLL Name: api-ms-win-crt-runtime-l1-1-0.dll
        DLL Name: api-ms-win-crt-heap-l1-1-0.dll
        DLL Name: api-ms-win-crt-string-l1-1-0.dll
        DLL Name: api-ms-win-crt-math-l1-1-0.dll
        DLL Name: VCRUNTIME140.dll
        DLL Name: python312.dll
        DLL Name: VCRUNTIME140_1.dll
        DLL Name: KERNEL32.dll
matplotlib/_path.cp312-win_amd64.pyd
        DLL
 vma:            Hint    Time      Forward  DLL       First
        DLL Name: api-ms-win-crt-runtime-l1-1-0.dll
        DLL Name: api-ms-win-crt-string-l1-1-0.dll
        DLL Name: api-ms-win-crt-math-l1-1-0.dll
        DLL Name: api-ms-win-crt-heap-l1-1-0.dll
        DLL Name: VCRUNTIME140.dll
        DLL Name: python312.dll
        DLL Name: VCRUNTIME140_1.dll
        DLL Name: KERNEL32.dll
matplotlib/_qhull.cp312-win_amd64.pyd
        DLL
 vma:            Hint    Time      Forward  DLL       First
        DLL Name: api-ms-win-crt-runtime-l1-1-0.dll
        DLL Name: api-ms-win-crt-stdio-l1-1-0.dll
        DLL Name: api-ms-win-crt-heap-l1-1-0.dll
        DLL Name: api-ms-win-crt-string-l1-1-0.dll
        DLL Name: api-ms-win-crt-math-l1-1-0.dll
        DLL Name: VCRUNTIME140.dll
        DLL Name: python312.dll
        DLL Name: api-ms-win-crt-utility-l1-1-0.dll
        DLL Name: api-ms-win-crt-time-l1-1-0.dll
        DLL Name: api-ms-win-crt-convert-l1-1-0.dll
        DLL Name: VCRUNTIME140_1.dll
        DLL Name: KERNEL32.dll
matplotlib/_tri.cp312-win_amd64.pyd
        DLL
 vma:            Hint    Time      Forward  DLL       First
        DLL Name: api-ms-win-crt-runtime-l1-1-0.dll
        DLL Name: api-ms-win-crt-stdio-l1-1-0.dll
        DLL Name: api-ms-win-crt-heap-l1-1-0.dll
        DLL Name: api-ms-win-crt-string-l1-1-0.dll
        DLL Name: api-ms-win-crt-math-l1-1-0.dll
        DLL Name: api-ms-win-crt-locale-l1-1-0.dll
        DLL Name: VCRUNTIME140.dll
        DLL Name: python312.dll
        DLL Name: api-ms-win-crt-filesystem-l1-1-0.dll
        DLL Name: VCRUNTIME140_1.dll
        DLL Name: KERNEL32.dll
matplotlib/_ttconv.cp312-win_amd64.pyd
        DLL
 vma:            Hint    Time      Forward  DLL       First
        DLL Name: api-ms-win-crt-runtime-l1-1-0.dll
        DLL Name: api-ms-win-crt-heap-l1-1-0.dll
        DLL Name: api-ms-win-crt-string-l1-1-0.dll
        DLL Name: api-ms-win-crt-math-l1-1-0.dll
        DLL Name: VCRUNTIME140.dll
        DLL Name: python312.dll
        DLL Name: api-ms-win-crt-stdio-l1-1-0.dll
        DLL Name: VCRUNTIME140_1.dll
        DLL Name: KERNEL32.dll
```

</details>

Apologies if this counts as a duplicate of #168 or #181 (or the others), but the messaging in those issues seemed... muddy.

**Comment by MatthieuDartiailh** [comment]
Thanks for the separate ticket which makes things clearer on my end. I will do my best to investigate ASAP.

**Comment by Qix-** [comment]
Another alternative is statically linking the runtimes. Then none of the redist DLLs would be necessary to begin with, instead of needing to package them alongside the plugin.

**Comment by MatthieuDartiailh** [comment]
Python ships VCRUNTIME so there is no problem there but MSVCP is problematic.

@QuLogic I would welcome your insight since I mostly followed your suggestions here. One difference I noticed between kiwi setup and Matplotlib is that in kiwi I re-enabled FH4, do you think it may be related ?

**Comment by QuLogic** [comment]
I see you copied the setup from Matplotlib here:
https://github.com/nucleic/kiwi/blob/16735f99f8d56c973c66c4b1522166ba215e015a/.github/workflows/release.yml#L106-L114
These are settings for Meson(-python) but I just noticed you are using `setuptools` to build. Those flags should make MSVCP statically linked with Meson, but they actually do nothing with `setuptools` (or perhaps something else entirely, as it didn't error?) Maybe there was a warning that could be seen if the build was in verbose mode?

From [`setuptools` docs](https://setuptools.pypa.io/en/stable/userguide/ext_modules.html#compiler-and-linker-options), it looks like you could set some environment variables to override build flags. The equivalent should be something like:
```
CFLAGS="/MT"
CPPFLAGS="/MT"
LDFLAGS="ucrt.lib vcruntime.lib /nodefaultlib:libucrt.lib /nodefaultlib:libvcruntime.lib"
```

**Comment by MatthieuDartiailh** [comment]
Thanks @QuLogic I missed this point. I considered switching to meson but never made the jump.

**Comment by MatthieuDartiailh** [comment]
I will have to try harder since my first attempt did not produce the expected result.

**Comment by QuLogic** [comment]
Maybe try [`CIBW_ENVIRONMENT_WINDOWS`](https://cibuildwheel.pypa.io/en/stable/options/#environment)

**Comment by MatthieuDartiailh** [comment]
Thanks for the suggestion but it does not seem to do the trick... (https://github.com/nucleic/kiwi/actions/runs/11999734728/job/33447988060) 
The logs looks like the env vars are set but still something is not behaving properly. I will investigate directly patching setuptools in cppy. I will have the added benefit of applying to all the projects that uses it.

**Comment by MatthieuDartiailh** [comment]
Building locally after patching cppy, I obtain the following dependencies, are those fine;
```
        DLL Name: api-ms-win-crt-runtime-l1-1-0.dll
        DLL Name: api-ms-win-crt-heap-l1-1-0.dll
        DLL Name: api-ms-win-crt-string-l1-1-0.dll
        DLL Name: api-ms-win-crt-stdio-l1-1-0.dll
        DLL Name: api-ms-win-crt-locale-l1-1-0.dll
        DLL Name: api-ms-win-crt-math-l1-1-0.dll
        DLL Name: VCRUNTIME140_1.dll
        DLL Name: VCRUNTIME140.dll
        DLL Name: python313.dll
        DLL Name: KERNEL32.dll
```

**Comment by MatthieuDartiailh** [comment]
Based on the first post of this issue I think so but I would be happy if somebody can confirm.

**Comment by LordAro** [comment]
Yup, vcruntime*.dll is distributed with cpython, kernel32.dll & api-*.dll are internal windows libraries, and python313 is... python :)

**Comment by QuLogic** [comment]
This is consistent with what we get: https://github.com/matplotlib/matplotlib/pull/28687#issuecomment-2278883165

**Comment by MatthieuDartiailh** [comment]
Thanks

**Comment by MatthieuDartiailh** [comment]
Can somebody more knowledgeable than myself in cibuildwheel config help me figure out the reason this pipeline fails ?

https://github.com/nucleic/kiwi/actions/runs/12036490541/job/33557857394

The override of MD with MT while not ideal does not cause issues locally nor in the CI test build.

**Comment by QuLogic** [comment]
Ah, that's a 32-bit build; we don't have those any more in Matplotlib. Perhaps slightly different library names are needed there and @zooba might know.

**Comment by MatthieuDartiailh** [comment]
Thanks for the pointer. Given the name disabling FH4 might help. I will try to give that a try to see if I can keep 32bits windows this way.

**Comment by MatthieuDartiailh** [comment]
Sadly it does not work... I will just drop 32 bits build for now and re-investigate if somebody complains about their absence.

**Comment by zooba** [comment]
> I will just drop 32 bits build for now and re-investigate if somebody complains about their absence.

Or document it as "for 32-bit builds you must also install the latest VC redist from https://aka.ms/vcredist" and move on with your life :)

**Comment by LordAro** [comment]
Looks like this has been fixed by #189 ? Any chance of a new release containing that in the next couple of weeks? I have deadlines :)

**Comment by MatthieuDartiailh** [comment]
I was considering waiting on the free-threaded build to be complete before cutting a new release. I will do my best to make a new release with or without free-threaded python support before Christmas.

**Comment by MatthieuDartiailh** [comment]
Since Setuptools new release is taking more time than I hoped I released 1.4.8 which fixes the linking issue and add support for free threaded Python except on Windows.

**Comment by spevs** [comment]
> Sadly it does not work... I will just drop 32 bits build for now and re-investigate if somebody complains about their absence.

It's about 6 months later now but FYI at my workplace we noticed immediately, as we started having errors when installing our internal library (which includes matplotlib as a dependency) on all new machines since the beginning of the year. At the time I was too busy to really investigate or open an issue so I just "fixed" it on our end by limiting kiwisolver in our pyproject dependencies to an older version. This is still working ok now but long term it would be good to have a solution if possible. We use 32-bit python due to some internal tools which depend on 32-bit DLLs.

**Comment by MatthieuDartiailh** [comment]
@spevs thanks for reporting. I will try using @zooba trick, but I will need you to test the resulting wheel.

**Comment by zooba** [comment]
@MatthieuDartiailh Also consider just not statically linking _or_ distributing MSVCP140 in the 32-bit build. I'm sure @spevs is capable of using the VCRedist installer - there's a good chance something else in those internal tools is installing it already (it might be an old version, but that's something the occasional 32-bit end user can sort out if they need to).

**Comment by MatthieuDartiailh** [comment]
It may not be that simple since cppy (which setups the compilation option) tries to build statically and that fails on 32-bit windows. I will look into reverting to dynamic linking on 32-bits windows.

**Comment by zooba** [comment]
I think @spevs wanted it, I'm just trying to help resolve things without anyone having to do too much work.

**Comment by MatthieuDartiailh** [comment]
Sorry I did not go up enough in the discussion.
